homeassistant/components/screenlogic/config_flow.py

```diff
@@ -65,163 +79,153 @@ async def _async_has_devices(hass):
+        if entry.unique_id is not None
+    }
+
+
+async def async_get_mac_address(hass, ip_address, port):
+    """Connect to a screenlogic gateway and return the mac address."""
+    connected_socket = await hass.async_add_executor_job(
+        login.create_socket,
+        ip_address,
+        port,
+    )
+    if not connected_socket:
+        raise ScreenLogicError("Unknown socket error")
+    return await hass.async_add_executor_job(login.gateway_connect, connected_socket)
 
 
 class ScreenlogicConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Config flow to setup screen logic devices."""
 
     VERSION = 1
     CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
 
     def __init__(self):
         """Initialize ScreenLogic ConfigFlow."""
         self.discovered_gateways = {}
+        self.discovered_ip = None
 
     @staticmethod
     @callback
     def async_get_options_flow(config_entry):
         """Get the options flow for ScreenLogic."""
         return ScreenLogicOptionsFlowHandler(config_entry)
 
-    async def async_step_import(self, import_config):
-        """Import a config entry from configuration.yaml."""
-        return await self.async_step_gateway_entry(import_config)
-
     async def async_step_user(self, user_input=None):
         """Handle the start of the config flow."""
-        _LOGGER.debug("async_step_user: user_input")
-        _LOGGER.debug(user_input)
-
-        # First, attempt to discover a ScreenLogic Gateway
-        if not user_input:
-            _LOGGER.debug("No input: Discover")
-            hosts = []
-            hosts = await self.hass.async_add_executor_job(discover_gateways)
-            if len(hosts) > 0:
-                for host in hosts:
-                    self.discovered_gateways[host[SL_GATEWAY_IP]] = host
-                return await self.async_step_gateway_select()
-            else:
-                return await self.async_step_gateway_entry()
+        self.discovered_gateways = await async_discover_gateways_by_unique_id(self.hass)
+        return await self.async_step_gateway_select()
+
+    async def async_step_dhcp(self, dhcp_discovery):
+        """Handle dhcp discovery."""
+        mac = _extract_mac_from_name(dhcp_discovery[HOSTNAME])
+        await self.async_set_unique_id(mac)
+        self._abort_if_unique_id_configured(
+            updates={CONF_IP_ADDRESS: dhcp_discovery[IP_ADDRESS]}
+        )
+        self.discovered_ip = dhcp_discovery[IP_ADDRESS]
+        self.context["title_placeholders"] = {"name": dhcp_discovery[HOSTNAME]}
+        return await self.async_step_gateway_entry()
 
     async def async_step_gateway_select(self, user_input=None):
         """Handle the selection of a discovered ScreenLogic gateway."""
         _LOGGER.debug("Gateway Select")
         _LOGGER.debug(
             *[gateway[SL_GATEWAY_NAME] for gateway in self.discovered_gateways.values()]
         )
+        existing = configured_instances(self.hass)
```

Comment:
There is a built in for this but I figured I didn't want to make this change set any larger since it's already a bit more then I intended to send

Reply:
Curious as to what the built in for this is, as I had seen it done this way in other integrations.

Reply:
`configured_instances` can likely be dropped and replaced with `self._async_current_ids()`
https://github.com/home-assistant/core/blob/dev/homeassistant/config_entries.py#L1036

Reply:
Nice. Looks like that will be a drop in replacement. Thanks!